src/hotspot/share/gc/shared/memAllocator.cpp

```diff
@@ -84,7 +83,29 @@ class MemAllocator::Allocation: StackObj {
   ~Allocation() {
     if (!check_out_of_memory()) {
       verify_after();
-      notify_allocation(_thread);
+
+      // Real allocation allocates actual memory: either allocates
+      // outside the TLAB, or allocates a new TLAB.
+      const bool is_real_allocation = _allocated_outside_tlab ||
+                                      (_allocated_tlab_size != 0);
+
+      if (is_real_allocation) {
+        notify_allocation_low_memory_detector();
+        notify_allocation_jfr_sampler();
+      }
+
+      if ((is_real_allocation || _tlab_end_reset_for_sample) &&
+          JvmtiExport::should_post_sampled_object_alloc()) {
```

Comment:
The same check is done inside the caller already.

On this note, I think all checks using info outside `class Allocation` should not be present on this level. (Ofc, this is very subjective.)

Reply:
This is a fast-path code, and it is extremely sensitive to the overheads of the actual calls. So I made a point to check the pre-conditions before calling them.

Reply:
> the overheads of the actual calls

I was expecting the same perf, since the compiler can see everything in this scope.

After hiding checks from the this level, I see the same number of `call` instructions in `MemAllocator::allocate`, where everything is inlined into. Ofc, still needs full perf-eval.